esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,289 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+            context.setAuthStatus(AUTH_STATUS_NOACCESS);
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate - Permission denied.");
+        }
+        context.ensureFeatureAccess(FEATURE_URL, accessType, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate: Permission denied.");
+
+        const char *fileName = requests.getName();
+        const char *clusterName = requests.getCluster();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(fileName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Name defined.");
+        if (isEmptyString(clusterName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Cluster defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        StringBuffer groupName;
+        if (!createWsDFUCreateGroupName(req.getPartLocations(), groupName))
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in createWsDFUCreateGroupName().");
+
+        Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName);
+        ClusterType clusterType = clusterInfo->getPlatform();
+        if (clusterType == RoxieCluster)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Roxie cluster is not allowed: %s.", clusterName);
+
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+
+        StringBuffer userId;
+        Owned<IUserDescriptor> userDesc;
+        context.getUserID(userId);
+        if (!userId.isEmpty())
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userId.str(), context.queryPassword(), context.querySignature());
+        }
+
+        StringBuffer tempFileName = fileName;
+        tempFileName.append(".wsdfucreate.tmp");
+
+        StringBuffer jobId = requests.getJobId();
+        if (jobId.isEmpty())
+            jobId.appendf("Create %s on %s", tempFileName.str(), clusterName);
+        ESPLOG(LogMax, "%s", jobId.str());
+        Owned<IFileDescriptor> fileDesc = createDFUDescriptor(context, tempFileName.str(),
+            clusterType, jobId.str(), userId.str(), req.getPartLocations(), groupName);
+        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fileDesc);
+        file->setAccessed();
+        file->setECL(recordDefinition);
+        file->queryAttributes().setPropBin("_rtlType", layoutBin.length(), layoutBin.toByteArray());
+        file->queryAttributes().setPropInt64("@recordCount", 0);
+        file->queryAttributes().setPropInt64("@size", 0);
+
+        getFileAccessBeforePublish(context, tempFileName.str(), jobId.str(), layoutBin, fileDesc, userDesc, requests, resp);
+
+        //create FileId
+        groupName.append(DFUFileIdSeparator).append(clusterName).append(DFUFileIdSeparator).append(tempFileName.str());
```

Comment:
for clarify I'd use a new variable here i.e. VStringBuffer fileID;
and this could be the point you encode in the record definition.
Something to think about :The whole thing could be compressed and base64'd.. )

Reply:
Do you want to base64 now?

Reply:
no, I don't think it's worth it unless were compressing and encoding something else (e.g. the record def). - for now plain text is fine.